### ISSUE TITLE
How to ensure that the selection coefficient satisfy the constraints: a+b=1

### ISSUE THREAD

**Issue by SkyeLu** [ResearchingDexter/pytorch_SKNet#4]
Hi. In the part of selection, the original paper wrotes that 

> In the case of two branches, the matrix B is redundant because ac + bc = 1.

but in your code: 
`       a_b=self.fc2(z)
        a_b=a_b.reshape(batch_size,self.M,self.out_channels,-1)
        a_b=self.softmax(a_b)
        #the part of selection
        a_b=list(a_b.chunk(self.M,dim=1))#split to a and b`
I did not see any constraints to a and b to ensure that a + b = 1.  Does this impact the performance of the network?
Thanks.

**Comment by MoonWolf9067** [comment]
I am also curious about this. @ResearchingDexter

**Comment by ResearchingDexter** [comment]
The softmax function is `self.softmax=nn.Softmax(dim=1)`, the `dim=1` is the dimension of `M` which ensures that **a+b=1**. @SkyeLu @MoonWolf9067

**Comment by MoonWolf9067** [comment]
I got it, thank you very much!
![Screenshot from 2019-06-02 15-02-37](https://user-images.githubusercontent.com/22995934/58758014-ce30fd00-8547-11e9-87c0-411272a396e0.png)
the code above is to reshape a_b to [batch_size, channel , height, width],  right? @ResearchingDexter

**Comment by ResearchingDexter** [comment]
The code `a_b=a_b.reshape(batch_size,self.M,self.out_channels,-1)` is to reshape `a_b` to `[batch_size, M, out_channel, 1]`. The `a_b` is `U->global_pool->fc1->fc2` @MoonWolf9067

**Comment by MoonWolf9067** [comment]
Thank you very much for your reply! In your code: **a_b=self.softmax(a_b)**, the softmax operates on the **channel dimension** and if I want to rewrite **a_b=a_b.reshape(batch_size, self.M, self.out_channels, -1**) with **tensorflow**, it should be: **a_b=a_b.reshape(batch_size,  self.out_channels, -1, self.M)**, right? @ResearchingDexter

**Comment by ResearchingDexter** [comment]
I'm not sure whether the channel dimension is the last in the tensorflow. If it is, I think it should be `a_b=a_b.reshape(batch_size, -1, self.out_channels, self.M)` or `a_b=a_b.reshape(batch_size, -1, self.M, self.out_channels)` with tensorflow. And remember the **softmax** and **split** operate on the `M` dimension. @MoonWolf9067

**Comment by MoonWolf9067** [comment]
Thank you very much! I know what you said and the shape of the feature map in **tensorflow** is:**[Batch_size, H, W, Channel]**, while, it is:**[Batch_size, Channel, H, W]** in **pytorch.** @ResearchingDexter

**Comment by ResearchingDexter** [comment]
OK, I got it. Thank you! @MoonWolf9067
